File: metacasa-ios/MetaCasa/Models/RecurringTransaction.swift
import Foundation
import SwiftUI

struct RecurringTransaction: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    var householdId: UUID
    var userId: UUID
    var type: TxType
    var amount: Decimal
    var category: String
    var subcategory: String?
    var account: String?
    var startDate: Date
    var endDate: Date?
    var nextDate: Date?
    var note: String?
    var frequency: Frequency
    var active: Bool
    let createdAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case householdId = "household_id"
        case userId = "user_id"
        case type
        case amount
        case category
        case subcategory
        case account
        case startDate = "start_date"
        case endDate = "end_date"
        case nextDate = "next_date"
        case note
        case frequency
        case active
        case createdAt = "created_at"
    }
}

enum Frequency: String, Codable, Hashable, Sendable, CaseIterable {
    case daily, weekly, monthly, yearly

    /// LocalizedStringKey para usar directamente en Text en SwiftUI.
    var labelKey: LocalizedStringKey {
        switch self {
        case .daily: "freq.daily"
        case .weekly: "freq.weekly"
        case .monthly: "freq.monthly"
        case .yearly: "freq.yearly"
        }
    }

    var label: String {
        switch self {
        case .daily: String(localized: "freq.daily")
        case .weekly: String(localized: "freq.weekly")
        case .monthly: String(localized: "freq.monthly")
        case .yearly: String(localized: "freq.yearly")
        }
    }

    var systemIcon: String {
        switch self {
        case .daily: "calendar"
        case .weekly: "calendar.day.timeline.left"
        case .monthly: "calendar.badge.clock"
        case .yearly: "calendar.circle.fill"
        }
    }

    func nextDate(from date: Date) -> Date? {
        let cal = Calendar.current
        switch self {
        case .daily:   return cal.date(byAdding: .day,   value: 1,  to: date)
        case .weekly:  return cal.date(byAdding: .day,   value: 7,  to: date)
        case .monthly: return cal.date(byAdding: .month, value: 1,  to: date)
        case .yearly:  return cal.date(byAdding: .year,  value: 1,  to: date)
        }
    }
}

File: metacasa-flutter/lib/models/recurring_transaction.dart
import 'package:decimal/decimal.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

import '../core/utils/decimal_converter.dart';
import 'transaction.dart';

part 'recurring_transaction.freezed.dart';
part 'recurring_transaction.g.dart';

/// Frecuencia de recurrencia.
enum Frequency {
  @JsonValue('daily')
  daily,
  @JsonValue('weekly')
  weekly,
  @JsonValue('monthly')
  monthly,
  @JsonValue('yearly')
  yearly;

  /// Próxima fecha a partir de [date], según la frecuencia.
  /// Port de `Frequency.nextDate(from:)` en iOS.
  DateTime nextDate(DateTime date) => switch (this) {
        Frequency.daily => date.add(const Duration(days: 1)),
        Frequency.weekly => date.add(const Duration(days: 7)),
        Frequency.monthly => DateTime(date.year, date.month + 1, date.day),
        Frequency.yearly => DateTime(date.year + 1, date.month, date.day),
      };
}

/// Transacción recurrente (plantilla que genera movimientos automáticos).
/// Espejo 1:1 de `public.recurring_transactions`.
@freezed
abstract class RecurringTransaction with _$RecurringTransaction {
  const factory RecurringTransaction({
    required String id,
    @JsonKey(name: 'household_id') required String householdId,
    @JsonKey(name: 'user_id') required String userId,
    required TxType type,
    @DecimalConverter() required Decimal amount,
    required String category,
    String? subcategory,
    String? account,
    @JsonKey(name: 'start_date') required DateTime startDate,
    @JsonKey(name: 'end_date') DateTime? endDate,
    @JsonKey(name: 'next_date') DateTime? nextDate,
    String? note,
    required Frequency frequency,
    required bool active,
    @JsonKey(name: 'created_at') DateTime? createdAt,
  }) = _RecurringTransaction;

  factory RecurringTransaction.fromJson(Map<String, dynamic> json) =>
      _$RecurringTransactionFromJson(json);
}
